src/Microsoft.ML.CpuMath/AvxIntrinsics.cs

```diff
@@ -268,27 +358,30 @@ public static unsafe void MatMulX(bool add, AlignedArray mat, AlignedArray src,
 
         public static unsafe void MatMulTranX(bool add, AlignedArray mat, AlignedArray src, AlignedArray dst, int crow, int ccol)
         {
-            Contracts.Assert(HasCompatibleAlignment(mat));
-            Contracts.Assert(HasCompatibleAlignment(src));
-            Contracts.Assert(HasCompatibleAlignment(dst));
+            Contracts.Assert(crow % 4 == 0);
+            Contracts.Assert(ccol % 4 == 0);
 
-            fixed (float* pSrcStart = &src.Items[0])
-            fixed (float* pDstStart = &dst.Items[0])
-            fixed (float* pMatStart = &mat.Items[0])
-            {
-                float* psrc = GetAlignedBase(src, pSrcStart);
-                float* pdst = GetAlignedBase(dst, pDstStart);
-                float* pmat = GetAlignedBase(mat, pMatStart);
+            MatMulTranX(add, mat.Items, src.Items, dst.Items, crow, ccol);
+        }
 
+        public static unsafe void MatMulTranX(bool add, float[] mat, float[] src, float[] dst, int crow, int ccol)
+        {
+            fixed (float* psrc = &src[0])
+            fixed (float* pdst = &dst[0])
+            fixed (float* pmat = &mat[0])
+            fixed (uint* pLeadingAlignmentMask = &LeadingAlignmentMask[0])
+            fixed (uint* pTrailingAlignmentMask = &TrailingAlignmentMask[0])
+            {
                 float* pSrcEnd = psrc + ccol;
                 float* pDstEnd = pdst + crow;
                 float* pSrcCurrent = psrc;
                 float* pMatCurrent = pmat;
+                bool firstTime = true;
```

Comment:
This is hurting perf, because it requires a branch check in the hot path of the code.

Reply:
I am working on removing the add flag as we are not using it anywhere. hopefully after that it will get simplified little bit more.
Otherwise we can always move back to doing the first iteration outside the loop

Reply:
I have removed the add flag. Now its like first operation is store and after that it is ```add + store```.
If we clear destination array first, it will avoid duplication of the code and we wont need the ```firstTime``` Boolean variable.
Another would be to move common code to a function.
what would be the best way to move ahead ?